Invalid module name: 'native' is not a Java identifier

Could you please add Automatic-Module-Name to manifest? Thanx;).

I'm running in the same issue.
This happens when trying to load a brotli4j native (e.g. native-linux-x86_64) in a modular context.
The JMPS needs to assign a module name to the jar, and since no module-info is present, it tries to give the jar a generated name. Here is where the problem happens: Module names can contain restricted keywords (static, abstract, native, public, etc), but because the jar starts with "native", the automatic name starts with "native" as well, which is not allowed and causes the JPMS to crash later.
You can see an example crash here:
```
Exception in thread "main" java.lang.module.FindException: Unable to derive module descriptor for /home/tobias/.gradle/caches/modules-2/files-2.1/com.aayushatharva.brotli4j/native-linux-x86_64/1.8.0/add74b11206e3a6ff14c7405718a0b61abae9a75/native-linux-x86_64-1.8.0.jar
	at java.base/jdk.internal.module.ModulePath.readJar(ModulePath.java:648)
	at java.base/jdk.internal.module.ModulePath.readModule(ModulePath.java:331)
	at java.base/jdk.internal.module.ModulePath.scan(ModulePath.java:237)
	at java.base/jdk.internal.module.ModulePath.scanNextEntry(ModulePath.java:190)
	at java.base/jdk.internal.module.ModulePath.find(ModulePath.java:154)
	at java.base/java.lang.module.Resolver.findWithBeforeFinder(Resolver.java:842)
	at java.base/java.lang.module.Resolver.resolve(Resolver.java:182)
	at java.base/java.lang.module.Resolver.resolve(Resolver.java:141)
	at java.base/java.lang.module.Configuration.resolveAndBind(Configuration.java:492)
	at java.base/java.lang.module.Configuration.resolveAndBind(Configuration.java:298)
Caused by: java.lang.IllegalArgumentException: native.linux.x86.64: Invalid module name: 'native' is not a Java identifier
	at java.base/jdk.internal.module.Checks.requireModuleName(Checks.java:52)
	at java.base/java.lang.module.ModuleDescriptor$Builder.<init>(ModuleDescriptor.java:1515)
	at java.base/java.lang.module.ModuleDescriptor.newAutomaticModule(ModuleDescriptor.java:2395)
	at java.base/jdk.internal.module.ModulePath.deriveModuleDescriptor(ModulePath.java:508)
	at java.base/jdk.internal.module.ModulePath.readJar(ModulePath.java:644)
	... 10 more
```
The authors of the java module system put a system in place that allows you to work around this issue without the need of including full module support via module-info: The Automatic-Module-Name.
The Automatic-Module-Name is a property read from the manifest that determines the name of the automatic module when loading in a modular context.
See https://github.com/jaxen-xpath/jaxen/commit/ecf428224e32881f845663439629c28f949248d1 for an example of how to do this with maven.

--- a/brotli4j/pom.xml
+++ b/brotli4j/pom.xml
@@ -38,7 +38,7 @@
             <plugin>
                 <groupId>org.moditect</groupId>
                 <artifactId>moditect-maven-plugin</artifactId>
-                <version>1.0.0.RC1</version>
+                <version>1.0.0.RC2</version>
                 <executions>
                     <execution>
                         <id>add-module-infos</id>

--- a/brotli4j/pom.xml
+++ b/brotli4j/pom.xml
@@ -29,6 +29,47 @@
     <artifactId>brotli4j</artifactId>
     <packaging>jar</packaging>
 
+    <properties>
+        <javaModuleName>com.aayushatharva.brotli4j</javaModuleName>
+    </properties>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.moditect</groupId>
+                <artifactId>moditect-maven-plugin</artifactId>
+                <version>1.0.0.RC1</version>
+                <executions>
+                    <execution>
+                        <id>add-module-infos</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>add-module-info</goal>
+                        </goals>
+                        <configuration>
+                            <jvmVersion>9</jvmVersion>
+                            <module>
+                                <moduleInfo>
+                                    <name>${javaModuleName}</name>
+                                    <exports>
+                                        com.aayushatharva.brotli4j;
+                                        com.aayushatharva.brotli4j.common;
+                                        com.aayushatharva.brotli4j.decoder;
+                                        com.aayushatharva.brotli4j.encoder;
+                                    </exports>
+                                </moduleInfo>
+                            </module>
+                            <jdepsExtraArgs>
+                                <arg>--multi-release</arg>
+                                <arg>9</arg>
+                            </jdepsExtraArgs>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
     <profiles>
         <profile>
             <id>linux-x86_64</id>

--- a/brotli4j/src/main/java/com/aayushatharva/brotli4j/Brotli4jLoader.java
+++ b/brotli4j/src/main/java/com/aayushatharva/brotli4j/Brotli4jLoader.java
@@ -38,7 +38,8 @@ public class Brotli4jLoader {
         } catch (Throwable t) {
             try {
                 String nativeLibName = System.mapLibraryName("brotli");
-                String libPath = "/lib/" + getPlatform() + "/" + nativeLibName;
+                String platform = getPlatform();
+                String libPath = "/lib/" + platform + "/" + nativeLibName;
 
                 File tempDir = new File(System.getProperty("java.io.tmpdir"), "com_aayushatharva_brotli4j_" + System.nanoTime());
                 tempDir.mkdir();
@@ -46,7 +47,8 @@ public class Brotli4jLoader {
 
                 File tempFile = new File(tempDir, nativeLibName);
 
-                try (InputStream in = Brotli4jLoader.class.getResourceAsStream(libPath)) {
+                Class<?> loaderClassToUse = Class.forName("com.aayushatharva.brotli4j." + platform.replace('-', '.') + ".NativeLoader");
+                try (InputStream in = loaderClassToUse.getResourceAsStream(libPath)) {
                     Files.copy(in, tempFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
                 } catch (Throwable throwable) {
                     tempFile.delete();

--- a/natives/linux-aarch64/pom.xml
+++ b/natives/linux-aarch64/pom.xml
@@ -38,7 +38,7 @@
             <plugin>
                 <groupId>org.moditect</groupId>
                 <artifactId>moditect-maven-plugin</artifactId>
-                <version>1.0.0.RC1</version>
+                <version>1.0.0.RC2</version>
                 <executions>
                     <execution>
                         <id>add-module-infos</id>
@@ -51,6 +51,9 @@
                             <module>
                                 <moduleInfo>
                                     <name>${javaModuleName}</name>
+                                    <exports>
+                                        ${javaModuleName} to com.aayushatharva.brotli4j;
+                                    </exports>
                                 </moduleInfo>
                             </module>
                             <jdepsExtraArgs>

--- a/natives/linux-aarch64/pom.xml
+++ b/natives/linux-aarch64/pom.xml
@@ -29,6 +29,41 @@
     <artifactId>native-linux-aarch64</artifactId>
     <packaging>jar</packaging>
 
+    <properties>
+        <javaModuleName>com.aayushatharva.brotli4j.linux.aarch64</javaModuleName>
+    </properties>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.moditect</groupId>
+                <artifactId>moditect-maven-plugin</artifactId>
+                <version>1.0.0.RC1</version>
+                <executions>
+                    <execution>
+                        <id>add-module-infos</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>add-module-info</goal>
+                        </goals>
+                        <configuration>
+                            <jvmVersion>9</jvmVersion>
+                            <module>
+                                <moduleInfo>
+                                    <name>${javaModuleName}</name>
+                                </moduleInfo>
+                            </module>
+                            <jdepsExtraArgs>
+                                <arg>--multi-release</arg>
+                                <arg>9</arg>
+                            </jdepsExtraArgs>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
     <profiles>
         <profile>
             <id>linux-aarch64</id>

--- a/natives/linux-aarch64/src/main/java/com/aayushatharva/brotli4j/linux/aarch64/NativeLoader.java
+++ b/natives/linux-aarch64/src/main/java/com/aayushatharva/brotli4j/linux/aarch64/NativeLoader.java
@@ -1,0 +1,7 @@
+package com.aayushatharva.brotli4j.linux.aarch64;
+
+/**
+ * Empty class to access the native lib in a JPMS context
+ */
+public class NativeLoader {
+}

--- a/natives/linux-x86_64/pom.xml
+++ b/natives/linux-x86_64/pom.xml
@@ -38,7 +38,7 @@
             <plugin>
                 <groupId>org.moditect</groupId>
                 <artifactId>moditect-maven-plugin</artifactId>
-                <version>1.0.0.RC1</version>
+                <version>1.0.0.RC2</version>
                 <executions>
                     <execution>
                         <id>add-module-infos</id>
@@ -51,6 +51,9 @@
                             <module>
                                 <moduleInfo>
                                     <name>${javaModuleName}</name>
+                                    <exports>
+                                        ${javaModuleName} to com.aayushatharva.brotli4j;
+                                    </exports>
                                 </moduleInfo>
                             </module>
                             <jdepsExtraArgs>

--- a/natives/linux-x86_64/pom.xml
+++ b/natives/linux-x86_64/pom.xml
@@ -29,6 +29,41 @@
     <artifactId>native-linux-x86_64</artifactId>
     <packaging>jar</packaging>
 
+    <properties>
+        <javaModuleName>com.aayushatharva.brotli4j.linux.x86_64</javaModuleName>
+    </properties>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.moditect</groupId>
+                <artifactId>moditect-maven-plugin</artifactId>
+                <version>1.0.0.RC1</version>
+                <executions>
+                    <execution>
+                        <id>add-module-infos</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>add-module-info</goal>
+                        </goals>
+                        <configuration>
+                            <jvmVersion>9</jvmVersion>
+                            <module>
+                                <moduleInfo>
+                                    <name>${javaModuleName}</name>
+                                </moduleInfo>
+                            </module>
+                            <jdepsExtraArgs>
+                                <arg>--multi-release</arg>
+                                <arg>9</arg>
+                            </jdepsExtraArgs>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
     <profiles>
         <profile>
             <id>linux-x86_64</id>

--- a/natives/linux-x86_64/src/main/java/com/aayushatharva/brotli4j/linux/x86_64/NativeLoader.java
+++ b/natives/linux-x86_64/src/main/java/com/aayushatharva/brotli4j/linux/x86_64/NativeLoader.java
@@ -1,0 +1,7 @@
+package com.aayushatharva.brotli4j.linux.x86_64;
+
+/**
+ * Empty class to access the native lib in a JPMS context
+ */
+public class NativeLoader {
+}

--- a/natives/osx-aarch64/pom.xml
+++ b/natives/osx-aarch64/pom.xml
@@ -29,6 +29,44 @@
     <artifactId>native-osx-aarch64</artifactId>
     <packaging>jar</packaging>
 
+    <properties>
+        <javaModuleName>com.aayushatharva.brotli4j.macos.aarch64</javaModuleName>
+    </properties>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.moditect</groupId>
+                <artifactId>moditect-maven-plugin</artifactId>
+                <version>1.0.0.RC2</version>
+                <executions>
+                    <execution>
+                        <id>add-module-infos</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>add-module-info</goal>
+                        </goals>
+                        <configuration>
+                            <jvmVersion>9</jvmVersion>
+                            <module>
+                                <moduleInfo>
+                                    <name>${javaModuleName}</name>
+                                    <exports>
+                                        ${javaModuleName} to com.aayushatharva.brotli4j;
+                                    </exports>
+                                </moduleInfo>
+                            </module>
+                            <jdepsExtraArgs>
+                                <arg>--multi-release</arg>
+                                <arg>9</arg>
+                            </jdepsExtraArgs>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
     <profiles>
 
         <profile>

--- a/natives/osx-aarch64/src/main/java/com/aayushatharva/brotli4j/macos/aarch64/NativeLoader.java
+++ b/natives/osx-aarch64/src/main/java/com/aayushatharva/brotli4j/macos/aarch64/NativeLoader.java
@@ -1,0 +1,7 @@
+package com.aayushatharva.brotli4j.macos.aarch64;
+
+/**
+ * Empty class to access the native lib in a JPMS context
+ */
+public class NativeLoader {
+}

--- a/natives/osx-x86_64/pom.xml
+++ b/natives/osx-x86_64/pom.xml
@@ -38,7 +38,7 @@
             <plugin>
                 <groupId>org.moditect</groupId>
                 <artifactId>moditect-maven-plugin</artifactId>
-                <version>1.0.0.RC1</version>
+                <version>1.0.0.RC2</version>
                 <executions>
                     <execution>
                         <id>add-module-infos</id>
@@ -51,6 +51,9 @@
                             <module>
                                 <moduleInfo>
                                     <name>${javaModuleName}</name>
+                                    <exports>
+                                        ${javaModuleName} to com.aayushatharva.brotli4j;
+                                    </exports>
                                 </moduleInfo>
                             </module>
                             <jdepsExtraArgs>

--- a/natives/osx-x86_64/pom.xml
+++ b/natives/osx-x86_64/pom.xml
@@ -29,6 +29,41 @@
     <artifactId>native-osx-x86_64</artifactId>
     <packaging>jar</packaging>
 
+    <properties>
+        <javaModuleName>com.aayushatharva.brotli4j.macos.x86_64</javaModuleName>
+    </properties>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.moditect</groupId>
+                <artifactId>moditect-maven-plugin</artifactId>
+                <version>1.0.0.RC1</version>
+                <executions>
+                    <execution>
+                        <id>add-module-infos</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>add-module-info</goal>
+                        </goals>
+                        <configuration>
+                            <jvmVersion>9</jvmVersion>
+                            <module>
+                                <moduleInfo>
+                                    <name>${javaModuleName}</name>
+                                </moduleInfo>
+                            </module>
+                            <jdepsExtraArgs>
+                                <arg>--multi-release</arg>
+                                <arg>9</arg>
+                            </jdepsExtraArgs>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
     <profiles>
 
         <profile>

--- a/natives/osx-x86_64/src/main/java/com/aayushatharva/brotli4j/macos/x86_64/NativeLoader.java
+++ b/natives/osx-x86_64/src/main/java/com/aayushatharva/brotli4j/macos/x86_64/NativeLoader.java
@@ -1,0 +1,7 @@
+package com.aayushatharva.brotli4j.macos.x86_64;
+
+/**
+ * Empty class to access the native lib in a JPMS context
+ */
+public class NativeLoader {
+}

--- a/natives/windows-x86_64/pom.xml
+++ b/natives/windows-x86_64/pom.xml
@@ -29,6 +29,41 @@
     <artifactId>native-windows-x86_64</artifactId>
     <packaging>jar</packaging>
 
+    <properties>
+        <javaModuleName>com.aayushatharva.brotli4j.windows.x86_64</javaModuleName>
+    </properties>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.moditect</groupId>
+                <artifactId>moditect-maven-plugin</artifactId>
+                <version>1.0.0.RC1</version>
+                <executions>
+                    <execution>
+                        <id>add-module-infos</id>
+                        <phase>package</phase>
+                        <goals>
+                            <goal>add-module-info</goal>
+                        </goals>
+                        <configuration>
+                            <jvmVersion>9</jvmVersion>
+                            <module>
+                                <moduleInfo>
+                                    <name>${javaModuleName}</name>
+                                </moduleInfo>
+                            </module>
+                            <jdepsExtraArgs>
+                                <arg>--multi-release</arg>
+                                <arg>9</arg>
+                            </jdepsExtraArgs>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+        </plugins>
+    </build>
+
     <profiles>
         <profile>
             <id>windows-x86_64</id>

--- a/natives/windows-x86_64/pom.xml
+++ b/natives/windows-x86_64/pom.xml
@@ -38,7 +38,7 @@
             <plugin>
                 <groupId>org.moditect</groupId>
                 <artifactId>moditect-maven-plugin</artifactId>
-                <version>1.0.0.RC1</version>
+                <version>1.0.0.RC2</version>
                 <executions>
                     <execution>
                         <id>add-module-infos</id>
@@ -51,6 +51,9 @@
                             <module>
                                 <moduleInfo>
                                     <name>${javaModuleName}</name>
+                                    <exports>
+                                        ${javaModuleName} to com.aayushatharva.brotli4j;
+                                    </exports>
                                 </moduleInfo>
                             </module>
                             <jdepsExtraArgs>

--- a/natives/windows-x86_64/src/main/java/com/aayushatharva/brotli4j/windows/x86_64/NativeLoader.java
+++ b/natives/windows-x86_64/src/main/java/com/aayushatharva/brotli4j/windows/x86_64/NativeLoader.java
@@ -1,0 +1,7 @@
+package com.aayushatharva.brotli4j.windows.x86_64;
+
+/**
+ * Empty class to access the native lib in a JPMS context
+ */
+public class NativeLoader {
+}

--- a/pom.xml
+++ b/pom.xml
@@ -105,6 +105,18 @@
 
     <build>
         <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-jar-plugin</artifactId>
+                <version>3.2.0</version>
+                <configuration>
+                    <archive>
+                        <manifestEntries>
+                            <Multi-Release>true</Multi-Release>
+                        </manifestEntries>
+                    </archive>
+                </configuration>
+            </plugin>
 
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>

